Interactive Tutorial: Trade in the Decentralized Exchange (#1396)
XRPLF/xrpl-dev-portal: content/tutorials/use-tokens/trade-in-the-decentralized-exchange.md

---
parent: use-tokens.html
blurb: Buy or sell fungible tokens for each other or for XRP in the decentralized exchange.
embed_xrpl_js: true
filters:
  - interactive_steps
  - include_code
labels:
  - Tokens
---
# Trade in the Decentralized Exchange

> This tutorial demonstrates how you can buy and sell tokens in the [decentralized exchange](decentralized-exchange.html).

Reviewer: ```suggestion
This tutorial demonstrates how you can buy and sell tokens in the [decentralized exchange](decentralized-exchange.html) (DEX).
```

## Prerequisites

<!-- Source for this specific tutorial's interactive bits: -->
<script src='https://cdn.jsdelivr.net/npm/bignumber.js@9.0.2/bignumber.min.js'></script>
<script type="application/javascript" src="assets/js/tutorials/trade-in-the-dex.js"></script>

This page provides JavaScript examples that use the [xrpl.js](https://js.xrpl.org/) library. See [Get Started Using JavaScript](get-started-using-javascript.html) for setup instructions.

Since JavaScript works in the web browser, you can read along and use the interactive steps without any setup.

## Steps
{% set n = cycler(* range(1,99)) %}

This tutorial demonstrates how to buy a fungible token in the decentralized exchange by selling XRP. (Other types of trades are possible, but selling a token, for example, requires you to have it first.) The example token used in this tutorial is as follows:

| Currency Code | Issuer | Notes |
|---|---|---|

> | TST | rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd | A test token pegged to XRP at a rate of approximately 10 XRP per 1 TST. The issuer has existing Offers on the XRP Ledger Testnet to buy and sell these tokens for  |

Reviewer 2: ```suggestion
| TST | rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd | A test token pegged to XRP at a rate of approximately 10 XRP per 1 TST. The issuer has existing Offers on the XRP Ledger Testnet to buy and sell these tokens.  |
```

### {{n.next()}}. Connect to Network

You must be connected to the network to submit transactions to it. Additionally, some languages (including JavaScript) require a high-precision number library for performing calculations on currency amounts you may encounter in the ledger. The following code shows how to connect to a public XRP Ledger Testnet server a supported [client library](client-libraries.html) with the appropriate dependencies.

<!-- MULTICODE_BLOCK_START -->

_JavaScript_

{{ include_code("_code-samples/trade-in-the-decentralized-exchange/js/base-with-bignumber.js", language="js") }}

<!-- MULTICODE_BLOCK_END -->

**Note:** The JavaScript code samples in this tutorial use the [`async`/`await` pattern](https://javascript.info/async-await). Since `await` needs to be used from within an `async` function, the remaining code samples are written to continue inside the `main()` function started here. You can also use Promise methods `.then()` and `.catch()` instead of `async`/`await` if you prefer.

For this tutorial, click the following button to connect:

{% include '_snippets/interactive-tutorials/connect-step.md' %}

### {{n.next()}}. Get Credentials

> To transact on the XRP Ledger, you need an address and secret key, and some XRP. For development purposes, you can get these on the [{{use_network}}](parallel-networks.html) using the following interface:

Reviewer: ```suggestion
To transact on the XRP Ledger, you need an address, a secret key, and some XRP. For development purposes, you can get these on the [{{use_network}}](parallel-networks.html) using the following interface:
```

{% include '_snippets/interactive-tutorials/generate-step.md' %}

> When you're [building production-ready software](production-readiness.html), you should use an existing account, and manage your keys using a [secure signing configuration](set-up-secure-signing.html). The following code shows how to instantiate a class with the wallet instance:

Reviewer: I'm not quite sure what the last sentence means: "The following code shows how to instantiate a class with the wallet instance".

Maybe better wording would be: "Below shows how you can create a Wallet object to track your account and keys:"

Author: Yeah, it was meant to say something like "instantiate an instance of the Wallet class" but something like what you wrote is better.

<!-- MULTICODE_BLOCK_END -->

You can use this interface to send the transaction specified by the amounts in the previous step:

{{ start_step("Send OfferCreate") }}
<button id="send-offercreate" class="btn btn-primary previous-steps-required" data-wait-step-name="Wait">Send OfferCreate</button>
<div class="loader collapse"><img class="throbber" src="assets/img/xrp-loader-96.png">Sending...</div>
<div class="output-area"></div>
{{ end_step() }}

### {{n.next()}}. Wait for Validation

Most transactions are accepted into the next ledger version after they're submitted, which means it may take 4-7 seconds for a transaction's outcome to be final. If the XRP Ledger is busy or poor network connectivity delays a transaction from being relayed throughout the network, a transaction may take longer to be confirmed. (For information on how to set an expiration for transactions, see [Reliable Transaction Submission](reliable-transaction-submission.html).)

{{ start_step("Wait") }}
{% include '_snippets/interactive-tutorials/wait-step.md' %}
{{ end_step() }}

### {{n.next()}}. Check Metadata

> You can use the validated transaction's [metadata](transaction-metadata.html) to determine exactly what it did. (Don't use metadata from tentative transaction results, because it may be different from the [final result](finality-of-results.html), especially when using the decentralized exchange.) In case of an OfferCreate transaction, likely results include:
> 
> - Some or all of the Offer may have been filled by matching with existing Offers in the ledger.
> - The unmatched remainder, if any, has been placed into the ledger to await new matching Offers.
> - Other bookkeeping may have occurred, such as removing expired or unfunded Offers that would have matched.

Reviewer: I could be wrong, but I think since you wrote `likely results include` the options should be in the present perfect tense and can omit the conditional phrases.

```suggestion
You can use the validated transaction's [metadata](transaction-metadata.html) to determine exactly what it did. (Don't use metadata from tentative transaction results, because it may be different from the [final result](finality-of-results.html), especially when using the decentralized exchange.) In case of an OfferCreate transaction, likely results include:

- Some or all of the Offer being filled by matching with existing Offers in the ledger.
- Any unmatched remainder being placed into the ledger to await new matching Offers.
- Other bookkeeping occurring, such as removing expired or unfunded Offers that would have matched.
```

Author: I think the original reads better here. (Also, not sure if this is just a typo on your part or if I'm mis-reading you, but "perfect" tense is for when something is _done._ You've changed it to present continuous instead.)
